src/main/java/org/jabref/gui/util/Theme.java

```diff
@@ -33,45 +46,164 @@
 
     public static final String BASE_CSS = "Base.css";
 
+    /**
+     * A size limit above which Theme will not attempt to keep a data-embedded URL in memory for the CSS.
+     *
+     * It's tolerable for CSS to exceed this limit; the functional benefit of the encoded CSS is in some edge
+     * case error handling. Specifically, having a reference to a data-embedded URL means that the Preview Viewer
+     * isn't impacted if the source CSS file is removed while the application is running.
+     *
+     * If the CSS is over this limit, then the user won't see any functional impact, as long as the file exists. Only if
+     * it becomes unavailable, might there be some impact. First, the Preview Viewer when created might not be themed.
+     * Second, there is a very small chance of uncaught exceptions. Theme makes a best effort to avoid this:
+     * it checks for CSS file existence before passing it to the Preview Viewer for theming. Still, as file existence
+     * checks are immediately out of date, it can't be perfectly ruled out.
+     *
+     * At the time of writing this comment:
+     *
+     * <ul>
+     * <li>src/main/java/org/jabref/gui/Base.css is 33k</li>
+     * <li>src/main/java/org/jabref/gui/Dark.css is 4k</li>
+     * <li>The dark custom theme in the Jabref documentation is 2k, see
+     * <a href="https://docs.jabref.org/advanced/custom-themes">Custom themes</a></li>
+     * </ul>
+     *
+     * So realistic custom themes will fit comfortably within 48k, even if they are modified copies of the base theme.
+     *
+     * Note that Base-64 encoding will increase the memory footprint of the URL by a third.
+     */
+    private static final int MAX_IN_MEMORY_CSS_LENGTH = 48000;
+
     private static final Logger LOGGER = LoggerFactory.getLogger(Theme.class);
 
     private final Type type;
-    private final Path pathToCss;
-    private final Optional<URL> additionalCssToLoad;
+
+    /* String, Path, and URL formats of the path to the css. These are determined at construction.
+     Path and URL are only set if they are relevant and valid (i.e. no illegal characters).
+     In general, use additionalCssToLoad(), to also ensure file existence checks are performed
+     */
+    private final String cssPathString;
+    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+    private final Optional<Path> cssPath;
```

Comment:
The `cssPath` variable is only used in the constructor (and in the `additionalCssToLoad` method, which is also only called in the constructor) if I see this correctly. Thus, I would propose to only keep the `cssPathString` and `cssUrl` variables on the class level.

Reply:
I didn't get to this one yet, but it's up next.

I agree I didn't quite like all these class level fields.

Only intention behind it, was to avoid duplicate logging when the path is not valid. File existence can change but for a given file system -- e.g. Windows file system, a path is either valid or not. So, it only needs checking once (when converting path string to a Path object). a realistic example would be trying to put a http:// URL into the box, on a Windows machine. I gave this a go as I was curious as to whether URLs would be accepted. It resulted in an uncaught exception, since on Windows ':' is treated as an indication of a drive letter. http not being a valid drive letter! Hence, catching the invalid path exception and logging it.

So to compare the options in brief:

- Path conversion in constructor and stored in class field: less logging but makes the class more complicated with an unnecessary field for LIGHT and DARK themes
- Path conversion into additionalCssToLoad(): will log invalid path each time a Preview Viewer is created, but avoids a field at class level that is only used for CUSTOM themes

I'm very much inclined to agree, that second option is overall better. But just wanted to run my reasoning past you for the current approach -- what do you think?

Reply:
Yes, I also like the second one better. In addition, one could add a check in the preference window that the path is correct.

Reply:
Then it's agreed, I will make the change to the Path handling.

I figure to add a separate enhancement issue and PR for user input validation in the preferences dialog, which will follow up on this fault fix nicely.

Reply:
This is now done.

The logging for invalid paths is still just in the constructor, as I realised it was better to convert the URL back to Path in additionalCssToLoad(). It is much cleaner now I think.

One last check I would like to do, is to run all the manual tests again with a linked build, as Gradle presents the dark theme as a file for both the test task and the run task.

I guess I can do this by following this guide: https://github.com/JabRef/jabref/blob/master/docs/advanced-reading/jpackage.md

Reply:
I am having a problem with the jlink step of building the binary for manual testing:

After installing wixtoolset I ran the following from the "Creating a binary and debug it" guide:

./gradlew -PprojVersion="5.0.50013" -PprojVersionInfo="5.0-ci.13--2020-03-05--c8e5924" jpackageImage

I got the following error:

> > Task :jlink FAILED
> The module name specified in 'application.mainModule' (null) has not the expected value (org.jabref).
> Error: Modules org.jabref.merged.module and javafx.web export package javafx.scene.web to module org.jabref

I thought I'd try to figure out the build from the CI pipeline instead, and so ran the jlinkZip task instead, but it fails in the same way.

I've tried JDK 14 and 15 with the same result.

Not sure where to go from here?

Reply:
The issue comes when you have a local jdk with javafx modules installed. Because jabref also brings into the modules, they are duplicate in the module path and therefore clash. For jlink locally you need to have a jdk without included javafx modules. 
You don't need to run this jlink stuff locally. You can run gradlew run for testing or let the CI build the packages. You can download the artifacts from the CI

Reply:
Aha! That makes a great deal of sense!

Only reason for testing this way, is Gradle presents the dark theme as a file under sec/main/resources rather than jrt URL. So I want to be sure everything still works with a real world binary.

Reply:
Thanks that did the trick! Manual tests repeated and all passed, when running via JabRef.bat on a jlink binary.